File: mobile/test/features/sources/chapter_label_test.dart
import 'package:aistudio_mobile/features/sources/utils/chapter_label.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('chapterLabel', () {
    test('leads with the canonical chapter number, title beneath', () {
      final label = chapterLabel(number: 134, title: 'The Culprit (7)');
      expect(label.primary, 'Chapter 134');
      expect(label.secondary, 'The Culprit (7)');
    });

    test('does not repeat the backend "Chapter N" fallback title', () {
      final label = chapterLabel(number: 133, title: 'Chapter 133');
      expect(label.primary, 'Chapter 133');
      expect(label.secondary, isNull);
    });

    test('formats whole numbers without a decimal point', () {
      expect(chapterLabel(number: 134.0, title: null).primary, 'Chapter 134');
      expect(chapterLabel(number: 10.5, title: null).primary, 'Chapter 10.5');
    });

    test('strips a redundant "Chapter N" prefix from titles', () {
      expect(
        chapterLabel(number: 12, title: 'Chapter 12: The Hunt').secondary,
        'The Hunt',
      );
      expect(
        chapterLabel(number: 12, title: 'chapter 12 - The Hunt').secondary,
        'The Hunt',
      );
      // A decimal continuation is a different chapter, not a redundant prefix.
      expect(
        chapterLabel(number: 12, title: 'Chapter 12.5 Special').secondary,
        'Chapter 12.5 Special',
      );
      expect(
        chapterLabel(number: 12, title: 'Chapter 125').secondary,
        'Chapter 125',
      );
    });

    test('treats empty titles as number-only chapters', () {
      expect(chapterLabel(number: 12, title: '').secondary, isNull);
      expect(chapterLabel(number: 12, title: '   ').secondary, isNull);
      expect(chapterLabel(number: 12, title: null).secondary, isNull);
    });

    test('falls back to the title only when the number is missing', () {
      final label = chapterLabel(number: null, title: 'Oneshot');
      expect(label.primary, 'Oneshot');
      expect(label.secondary, isNull);

      expect(chapterLabel(number: null, title: '').primary, 'Chapter');
    });
  });
}

File: frontend/src/features/sources/chapter-label.test.ts
import { describe, expect, it } from "vitest";
import { chapterLabel } from "./chapter-label";

describe("chapterLabel", () => {
  it("leads with the canonical chapter number and shows the title beneath", () => {
    expect(chapterLabel({ number: 134, title: "The Culprit (7)" })).toEqual({
      primary: "Chapter 134",
      secondary: "The Culprit (7)",
    });
    expect(
      chapterLabel({ number: 135, title: "Heavenly Demon Inauguration (1)" }),
    ).toEqual({
      primary: "Chapter 135",
      secondary: "Heavenly Demon Inauguration (1)",
    });
  });

  it("does not repeat the backend's 'Chapter N' fallback title", () => {
    expect(chapterLabel({ number: 133, title: "Chapter 133" })).toEqual({
      primary: "Chapter 133",
      secondary: null,
    });
  });

  it("strips a redundant 'Chapter N' prefix from titles (e.g. MangaKatana)", () => {
    expect(chapterLabel({ number: 12, title: "Chapter 12: The Hunt" })).toEqual({
      primary: "Chapter 12",
      secondary: "The Hunt",
    });
    expect(chapterLabel({ number: 12, title: "chapter 12 - The Hunt" })).toEqual({
      primary: "Chapter 12",
      secondary: "The Hunt",
    });
    // A decimal continuation is a different chapter, not a redundant prefix.
    expect(chapterLabel({ number: 12, title: "Chapter 12.5 Special" })).toEqual({
      primary: "Chapter 12",
      secondary: "Chapter 12.5 Special",
    });
    expect(chapterLabel({ number: 12, title: "Chapter 125" })).toEqual({
      primary: "Chapter 12",
      secondary: "Chapter 125",
    });
  });

  it("strips a BARE ordinal prefix (Royal Road: '1. Good Morning Brother')", () => {
    expect(
      chapterLabel({ number: 1, title: "1. Good Morning Brother" }),
    ).toEqual({ primary: "Chapter 1", secondary: "Good Morning Brother" });
    expect(chapterLabel({ number: 12, title: "12 - The Hunt" })).toEqual({
      primary: "Chapter 12",
      secondary: "The Hunt",
    });
    expect(chapterLabel({ number: 12, title: "12) The Hunt" })).toEqual({
      primary: "Chapter 12",
      secondary: "The Hunt",
    });
    expect(chapterLabel({ number: 12.5, title: "12.5: Interlude" })).toEqual({
      primary: "Chapter 12.5",
      secondary: "Interlude",
    });
    expect(chapterLabel({ number: 7, title: "7." })).toEqual({
      primary: "Chapter 7",
      secondary: null,
    });
    expect(chapterLabel({ number: 7, title: "7" })).toEqual({
      primary: "Chapter 7",
      secondary: null,
    });
  });

  it("never mistakes another number for the chapter's own", () => {
    // The leading digits are part of the title, not an ordinal.
    expect(chapterLabel({ number: 1, title: "1000 Years of Solitude" })).toEqual({
      primary: "Chapter 1",
      secondary: "1000 Years of Solitude",
    });
    expect(chapterLabel({ number: 12, title: "125. The Hunt" })).toEqual({
      primary: "Chapter 12",
      secondary: "125. The Hunt",
    });
    // A decimal continuation is a different chapter — taking "12" and the dot
    // as a separator would leave the row reading "5 Special".
    expect(chapterLabel({ number: 12, title: "12.5 Special" })).toEqual({
      primary: "Chapter 12",
      secondary: "12.5 Special",
    });
  });

  it("treats empty titles as number-only chapters", () => {
    expect(chapterLabel({ number: 12, title: "" })).toEqual({
      primary: "Chapter 12",
      secondary: null,
    });
    expect(chapterLabel({ number: 12, title: "   " })).toEqual({
      primary: "Chapter 12",
      secondary: null,
    });
    expect(chapterLabel({ number: 12, title: null })).toEqual({
      primary: "Chapter 12",
      secondary: null,
    });
  });

  it("falls back to the title only when the number is missing", () => {
    expect(chapterLabel({ number: null, title: "Oneshot" })).toEqual({
      primary: "Oneshot",
      secondary: null,
    });
    expect(chapterLabel({ number: null, title: "" })).toEqual({
      primary: "Chapter",
      secondary: null,
    });
  });
});
